[READY]New Side Antagonist: Space Dragon

IndieanaJones opened the pull request:
## Changelog
:cl:
add: Space Dragon Mob
add: Space Dragon Antagonist Event
/:cl:
## About The Pull Request
This pull adds a space dragon mob (subsidiary of the Ash Drake) and a corresponding antagonist and event for it to spawn in.  The space dragon appears as a purple variation of the Ash Drake, but has a number of notable differences:
- Has 250 health
- Does 35 damage, between the lesser ash drake's 30 and the ash drake's 40
- Moves as fast as a space carp
- Does not shoot fire in a cone, but instead solely in a straight line (this fire goes a tad further than the ash drake's)
- No collision damage (moving into various objects won't damage them)
- No swooping attack, but instead has a tail sweep attack.  Alt-clicking will use it, and will knock all upright mobs down around the space dragon in a one-block radius.  The space dragon also spins around when it uses it as a visual indicator.  It also has a cooldown like the swoop, but it is slightly shorter (200 to 150)
- Shares the resistances of the Ash Drake and other megafauna, so it is resistant to lasers by 50%.

The mob spawns where carp are allowed to spawn, and is instructed on it's controls.  It's explained that it has been informed of whom the leader of these invaders into the space dragon's claim are, and the dragon is tasked with killing that person.   The person selected will always be one of the Heads of Staff, and if for some reason there are no heads of staffs available for choosing, a random crewmember will be chosen instead. When they spawn, a unique announcement also plays, alerting the crew that the dragon has spawned.

## Why It's Good For The Game
I firmly believe more variety in antags is always a good thing when it comes to this game, and people enjoy playing as an ash drake and fighting them, so having a variation of it as an antag should be add an enjoyable experience.  While it has potential to wreck parts of the station, a well-coordinated crew should be able to limit it's mobility due to it's low health pool, forcing it to play conservatively when the crew pressures it.

On another note, I'm completely open to any suggestions or corrections to this, as this is my first project working with SS13, with hopefully more to follow.

IndieanaJones committed:
Add files via upload

IndieanaJones committed:
Rename space_dragon.dm to code/modules/events/space_dragon.dm

IndieanaJones committed:
Adds Space Dragon

IndieanaJones committed:
Adds Space Dragon Event

IndieanaJones committed:
Remove Copied File

IndieanaJones committed:
Put antagonist definition in proper location for space dragon

Cyprex commented:
This sounds very powerful. Also it sounds like it would be very boring if played repeatedly due to the lack of objectives, other antags can pick up items and such.

wesoda25 commented:
I could see it being an admin spawn, but a natural spawn would not be good. What is the point?

Shadowflame909 commented:
How rare will the spawning of this be? Also at what time? Like for example at the hour mark space ninjas and Xenomorphs can spawn. Well I think xenomorphs can spawn earlier now, for some reason.

IndieanaJones committed:
Update space_dragon.dm

IndieanaJones commented:
Hey tralezab, thanks for leaving feedback!  Seeing your feedback is pretty valid, I could propose the following changes:
- Remove passive health regen, making it so the drake must gib corpses to refill health
- Add a "Kill x amount of people" requirement, so the drake must go to the station to fufill the requirement
- possibly buff its health a little to compensate for loss of health regen.

IndieanaJones commented:
Also, Shadowflame, it does indeed keep the mini-avatar feature of the ash drake to keep you from munching yourself.

Shadowflame909 commented:
Eh. What's the point of adding an event, if you'll never see it?

Laughter Demons are an event with a weight of 1. And I haven't seen them in 2 years of playing, ever spawn.

IndieanaJones commented:
The reason it's currently set to one is so when this antag is finalized, people could come to an agreement on how often it should spawn, and adjust the number accordingly.  Currently, if I add the notes I mentioned to tralezab, I believe we could give it a spawn weight of 8, or less being on par or less with the pirates.

Borisvanmemes commented:
Making it only happen after one full hour means you can bump up the weight a tad i'd say.

Shadowflame909 commented:
Yeah. Xenos can spawn 20 minutes in for some reason. They're a guaranteed round-ender.

An event as weak as this? Not resistant to laser damage like full ash-drakes, and no regeneration? The chance of it getting cheesed is high enough, that I think it warrants a bump in the chance of it spawning. If lame Benos can spawn 20 minutes in via the end, spawn on lava land via their ruin, and spawn on the station in xenobiology. I'm pretty sure this is nothing to worry about.

IndieanaJones commented:
> 
> 
> Yeah. Xenos can spawn 20 minutes in for some reason. They're a guaranteed round-ender.
> 
> An event as weak as this? Not resistant to laser damage like full ash-drakes, and no regeneration? The chance of it getting cheesed is high enough, that I think it warrants a bump in the chance of it spawning. If lame Benos can spawn 20 minutes in via the end, spawn on lava land via their ruin, and spawn on the station in xenobiology. I'm pretty sure this is nothing to worry about.

Mind you, what I listed is what I changed about the space dragon from the ash drake.  The space dragon has all the resistances of the ash drake, unlike the lesser ash drake.

imsxz commented:
seems like the type of antag that'll get cucked the very second anyone picks up a gun. not saying to buff it, but the idea in general strikes me that it'll be a typical mob that people salt about being overpowered up  until the point that it faces actual resistance. The gibbing part is kinda offputting too, but not that awful in the scope of things.

That's my two cents, I've not thought very hard about this, take my opinion with a grain of salt.

Shadowflame909 commented:
Yeah. IMO I think its spawn chance should either be raised, or it should be buffed a good bit. I would like to relate it to the Leaper. It's a giant Frog with 200 health, slow and can stun and gib people.

4 tiders with toolboxes can surround it and kill it in 1 minute.


That's the kind of scenario I see here.

IndieanaJones commented:
Considering it is player-contolled, its fire breath, and it's decent speed, I doubt 4 greytiders with toolboxes would stand much of a chance.  Those without fire resistance will need to put themselves out, which leaves them open to being bitten.  However, even after factoring in these elements, if anyone believes it'd be too weak, I'd be open to suggestions on buffing it.

IndieanaJones commented:
> You should make some sort of announcement when one spawns so that station has some time to prepare.

Already does so.  😃

IndieanaJones committed:
Removed passive health regen from space dragon

IndieanaJones committed:
Updated Weight and Earliness for Space Dragon

IndieanaJones committed:
Added Objective to Space Dragon, Updated Starting Blurb

IndieanaJones commented:
Updated the main post with updated information.  For those who already read it:
- Removed the passive hp regen
- Set the spawn weight and time to 8 at 30 minutes, making it as common as Space Pirates
- Now has an objective to kill a selected Head of Staff, if no Head of Staff are available, a random crewmember will be chosen in their stead.

IndieanaJones committed:
Cleaned Up the Objective Determiner

anconfuzedrock commented:
why does the laser resist even need to exist? 10 shots of 20 damage is weak enough already.

OverlyGeneric commented:
I don't think space fits well for a dragon. It works for carps because silly & being a fish.

Maybe make it a giant space whale and have the corpse be valuable so cargo can re-enact Moby Dick.

Other than that. 👍

IndieanaJones committed:
Actually fix the space dragon objectives

OverlyGeneric commented:
> Well, the sort of inspiration for the idea was Ridley from Metroid, and in that universe he fits in quite nicely. If I had any idea on how to add new sprites, I would make custom sprites which look more like him.

Just ask your local spriter in the /tg/ discord but even Ridley couldn't breathe in outerspace or breathe fire. It's probably better to be a space whale because it fits better with the whole space carp and giant space carp.

IndieanaJones commented:
Well, if others would rather prefer a space whale over a space dragon, they can use emotes and comment.  For now, until such want is expressed, it will stay as is.

IndieanaJones committed:
Fix the objective display for the space dragon's target

IndieanaJones commented:
At this point, unless other changes are pushed for, the pull request is finalized.

GuyonBroadway commented:
Lore suggestion: Some asshole wizard is responsible for a dragon on the station.

Dang wizards!

tralezab commented:
@Therapist1 but imagine if this was a CHINESE space dragon....

Tlaltecuhtli commented:
why a dragon in a space station

GuyonBroadway commented:
The wizards did it.

ExcessiveUseOfCobblestone commented:
It's the personification of the dragon constellation, draco.

Shadowflame909 commented:
When the wizard staff of changes a space carp

IndieanaJones committed:
Updated Space Dragon Intro Blurb

IndieanaJones committed:
Updated Centcom Message When Space Dragon Spawns

IndieanaJones commented:
Using the ideas mentioned here (along with the He-Man intro), I've rewritten the intro blurb to better explain the space dragon's existence.  I've also updated the Centcom message to be more identifiable as well.

ExcessiveUseOfCobblestone commented:
lol'd

GuyonBroadway commented:
We also now have a fish that turns into a dragon, it's pokemon reference too.

OverlyGeneric commented:
I feel like this is going to be PRed to be a wizard event rather than a normal event.

That's why I suggested a space whale. Just so this event doesn't get forced to be a wizard event. 

_**AAAAAAAAAAAAAAAAAAA**_

tralezab commented:
Is there some space whale sprite floating around i'm unaware of?

IndieanaJones commented:
> 
> 
> Is there some space whale sprite floating around i'm unaware of?

No, there isn't.  It'd have to made, though since the reasoning is to save the event from being made a wizard-only one, I'm hoping such a thing is unnecessary.

IndieanaJones committed:
Updated Space Dragon's description with new lore

IndieanaJones commented:
This is the last thing which needs to be changed for the current vision and implementation of the space dragon.  As of this moment, it is finished, if no more changes are necessary.

Memager commented:
did someone say    s  p  r  i  t  e  s

IndieanaJones commented:
If you did want to make sprites for this, I'd imagine the current version of this that fits with what people have said is like a purple gyradoes from pokemon, maybe with carp-like eyes.  Make sure it has tiny sprites too if you do it, that's an important thing for a larger mob to have.

tralezab commented:
![image](https://user-images.githubusercontent.com/40974010/52024788-e0762e00-24b6-11e9-85d0-bf75b7433836.png)
this is incomplete and super rough but would something like this be okay?

Whoneedspacee commented:
Damn if that was animated and looked a little more sharp I'd say that could be a boss of it's own haha.

IndieanaJones commented:
Hot damn, that looks awesome. Go with it.

IndieanaJones committed:
Clean Up Space Dragon's OnFire()

IndieanaJones commented:
Also, OnFire() was sloppy because it was the first change I made for the space dragon that worked, so now that I know precisely what I'm doing, I went in and fixed it up, removing all the unnecessary stuff.

tralezab commented:
oh yeah here's the dmi sorry, i did move it down a tiny bit because it was a little high up
![spacedragon](https://user-images.githubusercontent.com/40974010/52032541-33131280-24d6-11e9-9766-d4b9225554ab.png)

Whoneedspacee commented:
please animate the tail

IndieanaJones commented:
![in-gamespacedragon](https://user-images.githubusercontent.com/47086570/52035430-81390d80-24f8-11e9-97f3-1d9ea8c51eef.png)
I made very tiny adjustments, but this is how it looks in-game.

Shadowflame909 commented:
That's pretty badass

tralezab commented:
I moved it to the ground to prevent confusion on what tile it's on, but glad to see you're using it instead of a reused sprite

Memager commented:
add some outlines and some nice shading and you're good to go

IndieanaJones commented:
Alright, so update:
- Space dragon has a 2-part idle, it moves up and unfurls it's tail, along with it's mouth opening a bit more and eyes glowing a tad.  It looks really good, in my opinion.
- Outline and shading added.
- It's mini-sprite is now the space carp sprite (fitting with the backstory)
- sound effect added to the tail swipe
- Updated death message and sprite
Overall, this has turned out quite differently from how it started, but the process has been quite fun and something very interesting came out of it.  

Also, I don't think there's any issue with this becoming a wizard event, so unless it actually gets brought up by the code reviewer, I don't think it needs to be a concern.

IndieanaJones committed:
Add Space Dragon Sprites

IndieanaJones committed:
Sprite in wrong location

IndieanaJones committed:
Properly adds Space Dragon Sprites

IndieanaJones committed:
Adds Space Dragon Tiny Sprite to be Carp

IndieanaJones committed:
Space Dragon Uses New Sprites, Tiny Form Updated

IndieanaJones commented:
I didn't request a review though, so if you see this Shiz, it wasn't intentional.

IndieanaJones commented:
![in-gamespacedragon2](https://user-images.githubusercontent.com/47086570/52101979-03ced500-25ac-11e9-9b5b-c70bb5c05581.png)
Current sprite

PKPenguin321 commented:
Seems a lot smaller than I imagined

Shadowflame909 commented:
My problem with it becoming a wizard event is that, you'd never get to see it.

And when you do, it'd instantly get fucked over by the other wizard memes, that he'd be pulling.

Honestly its health aint much as long as you got a fire-extinguisher and a bat

IndieanaJones commented:
What I meant to say is I don't think anyone is going to force this into becoming one.  I'd be rather upset if it did become a wizard event, because you'd never see it.

IndieanaJones commented:
Also, I did try messing around with the size using transform, but only 1x and 2x look good.  The reason being is when it's anything in between, the eyes and teeth do some crazy things.  Besides, it's 64x64 sprite, same as an ash drake, it just doesn't take up as much space.

Therapist1 commented:
I think you only need to wait, not sure the exact system with the maintainers but it does take a little bit of time.

GuyonBroadway commented:
You wait for oranges to get out of bed and silently close it with not explanation.

lmwevil commented:
if it doesn't have an objective to horde shiny things then i'll be sad

optimumtact commented on code/modules/events/space_dragon.dm:
```diff
@@ -0,0 +1,45 @@
+	typepath = /datum/round_event/ghost_role/space_dragon
+	max_occurrences = 1
+	weight = 8
+	earliest_start = 30 MINUTES
```
this is to early, I would pitch at around 1 hour 10 or 1 hour 20

IndieanaJones replied:
This is the same weight as the pirates, and even they aren't too often of a spawn.  Also, most rounds are much shorter than 80 minutes, usually hovering around 50-60 minutes.  But, if that's what's truly needed...

IndieanaJones committed:
Increased Time Til' Dragon Spawn

IndieanaJones commented:
Okay, change made.  Easy, right?

optimumtact commented:
It depends on the server you are playing on, also the pirate spawn is hampered by it's requirement for several ghosts

ExcessiveUseOfCobblestone commented:
Is there a reason the repulse tail sweep (edited subtype) wasn't used?

ExcessiveUseOfCobblestone reviewed (commented):
^^

IndieanaJones commented:
I'm unsure as to what you're referring to.  The tail sweep I have is most certainly used, unless you're referring to something else?

ExcessiveUseOfCobblestone commented:
https://github.com/tgstation/tgstation/blob/f4c273399fb158a71536d3aeb4bca46f6f551620/code/modules/spells/spell_types/wizard.dm#L308

ExcessiveUseOfCobblestone commented:
No worries, although I'd prefer you just build off the premade spell (it comes with an action button and easy2view counter so you know when you can reactivate it).

ExcessiveUseOfCobblestone reviewed (changes requested):
Formalizing my comment above

Kromgar commented:
You should give it the ability of the morph to eat shit so it has to fly around and gobble up shiny and expensive shit

IndieanaJones committed:
Drake's Tail Sweep Now Uses Prebuilt Spell

Thanks to Cobblestone for the head's up on this.

IndieanaJones commented:
Done.  Tail sweep now uses a custom subtype of the repulse spell.

ExcessiveUseOfCobblestone commented on code/modules/mob/living/simple_animal/hostile/megafauna/drake.dm:
```diff
@@ -491,4 +491,75 @@ Difficulty: Medium
+	. = ..()
+	smallsprite.Remove(src)
+
+/mob/living/simple_animal/hostile/megafauna/dragon/space_dragon/proc/fire_stream(var/atom/at = target)
```
does this work (argument)? I didn't know the target var was useful for PCs

IndieanaJones replied:
It does.

IndieanaJones replied:
The ash drake uses the same argument as well.

ExcessiveUseOfCobblestone replied:
nice
